### ISSUE TITLE
File definition for suppressions should ignore path separator

### ISSUE THREAD

**Issue by viliam-durina** [checkstyle/checkstyle#3744]
Currently, if we want the suppression rules to work on both Windows and Linux, we have to use `[\\/]` as the path separator. Our current configuration only works on Linux (https://github.com/hazelcast/hazelcast/blob/master/checkstyle/suppressions.xml). If we replace all path separators with `[\\/]`, the file becomes lot harder to read, have a look at your https://github.com/checkstyle/checkstyle/blob/master/config/suppressions.xml.

I propose, that there could be an option to automatically convert backslashes to forward slashes on Windows before matching the regex:

For example:

The file name `c:\projects\program\com\example\package\impl\Clazz.java` will be matched by rule:

    <suppress checks="JavadocMethod" files="/impl/"/>

I even believe, that most people won't mind if this is done by default.

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/config_filters.html#SuppressionFilter
> files - a Regular Expression matched against the file name associated with an audit event. It is mandatory. 

problem is that user will have non certain situation in what slashes to write config. Window users will do by windows way, unix people by their way. It will create just another problem and misunderstanding.
Yes, it is pain, we have it too. ...  I do not see good solution for now. Lets keep that issue as place for community to share ideas.
But good solution is not clear for me

**Comment by viliam-durina** [comment]
Java happily opens files on Windows, if they are written with Unix separators:

    // this will work on windows, even though windows is in uppercase and forward slash is used
    InputStream is = new FileInputStream("c:/windows/system.ini");

However, backslash will not work on Linux

    // this will throw java.io.FileNotFoundException on Linux
    InputStream is = new FileInputStream("\\etc\\mtab");

In Java, everyone is used to use forward slashes for paths, even though the system uses backslashes.

So I suggest the following algorithm:

    private static boolean fileMatches(String regexp, String file) {
        // first try to match to the original file name
        if (containsRegexp(regexp, file))
            return true;

        // if it did not match and we are on windows, replace the path separator character and retry
        if (File.separatorChar == '\\' && containsRegexp(regexp, file.replace('\\', '/')))
             return true;

        return false;
    }

    private static boolean containsRegexp(String regexp, String file) {
        return Pattern.compile(regexp).matcher(file).find();
    }

Here is a little test:

    private static final String[] files = { "c:\\work\\Program.java", "/work/Program.java"};
    private static final String[] regexps = { "[\\\\/]work[\\\\/]Program\\.java", 
        "/work/Program.java", "\\\\work\\\\Program.java"};

    public static void main(String[] args) throws IOException {
        for (String file : files)
            for (String regexp : regexps) {
                System.out.println("file=" + file);
                System.out.println("regexp=" + regexp);
                System.out.println("fileMatches=" + fileMatches(regexp, file));
                System.out.println("-----------");
            }
    }

And it's output:

		file=c:\work\Program.java
		regexp=[\\/]work[\\/]Program\.java
		fileMatches=true
		-----------
		file=c:\work\Program.java
		regexp=/work/Program.java
		fileMatches=true
		-----------
		file=c:\work\Program.java
		regexp=\\work\\Program.java
		fileMatches=true
		-----------
		file=/work/Program.java
		regexp=[\\/]work[\\/]Program\.java
		fileMatches=true
		-----------
		file=/work/Program.java
		regexp=/work/Program.java
		fileMatches=true
		-----------
		file=/work/Program.java
		regexp=\\work\\Program.java
		fileMatches=false
		-----------

This way we are backwards compatible.
- on Linux, only `'/'` and `'[\\/]'` work and `'\\'` does not (same as now)
- on Windows, all of `'\\'`, `'/'` and `'[\\/]'` work (currently, `'/'` does not)
- `'/'` and `'[\\/]'` work everywhere
- if one has a file with `'\'` on Linux in its name (this is allowed), it is not matched against `'/'`. However, such file is not transferrable to Windows.
- [one cannot use `'/'` in file name on Windows](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx), so no problem here.

If we want to be super-safe, we can add option to disable this feature.

**Comment by rnveach** [comment]
> In Java, everyone is used to use forward slashes for paths

I am a Windows user and I still use backslashes regardless. Most of the time it is because I am copying paths from explorer instead of writing them out by hand. I usually work only with Windows outside Checkstyle, so I've had little need to learn to use forward slashes for linux. Even with working on Checkstyle, the majority of my time is not spent working with linux or it's own caveats.

> Java happily opens files on Windows, if they are written with Unix separators:
> This way we are backwards compatible.
> 	file=/work/Program.java
	regexp=\\work\\Program.java
	fileMatches=false

If Checkstyle executed this on a linux machine and the suppression regexp was originally written on a windows machine, then this this example shows you are not fully compatible with each OS as `fileMatches` should be `true`. You would need another `if` and a conversion from `/` to `\`.

**Comment by viliam-durina** [comment]
If it's implemented according to my proposal, you'll be fine on Windows only.

	file=c:\work\Program.java
	regexp=\\work\\Program.java
	fileMatches=true

If you want to be portable, you'll just have to use `'/'` everywhere. That's it.

**Comment by romani** [comment]
idea of @viliam-durina need to be considered in details.

Another idea: is define filesystem marker in regexp string. Kind of similar idea as at gradle - https://docs.gradle.org/current/userguide/userguide_single.html#sec:executing_a_multiproject_build

> For more control over what gets executed, use qualified names (the second approach mentioned). These are paths just like directory paths, but use ‘:’ instead of ‘/’ or ‘\’. If the path begins with a ‘:’, then the path is resolved relative to the root project. In other words, the leading ‘:’ represents the root project itself. All other colons are path separators.


So we can use new option:
treatAsFileSystemSerator="\"
path="src\org\...."

Check will substitude `treatAsFileSystemSerator` in `path`  before RegExp is build by Filesystem separator of running OS. Whole responsibility on user for making this work.

**Comment by pbludov** [comment]
I fell into this trap today. After a short discussion, we came to
```
<suppressions>
  <!-- Common suppressions for all tests -->
  <suppress checks="(Javadoc|MissingCtor)" files="src.test.java"/>
```
which is less reliable but much more readable.
By the way, there is one more problem: windows ignores case in file names.
So, in addition to treatAsFileSystemSerator, we also need caseSensitive="true|false".

**Comment by romani** [comment]
@pbludov 

> we also need caseSensitive="true|false".

What problems do you experience ?  what is a benefit to be less strict ? 
it is better to create new issue for this discussion.

**Comment by pbludov** [comment]
I saw a problem like this in the Ant world. It was related to the generated files.
Perhaps you are right, and such a problem in Checkstyle will be unlikely.

Nevertheless, the 'glob' mapper from Ant has two attributes:

casesensitive
    If this is false, the mapper will ignore case when matching the glob pattern.

handledirsep
    If this is specified, the mapper will ignore the difference between the normal
    directory separator characters - \ and /.

See https://ant.apache.org/manual/Types/mapper.html

**Comment by snago** [comment]
I'd like to try to fix this.

What PR would be most likely to get accepted?
* Auto-replace `\\` to `/` in paths on Windows (like @viliam-durina suggested in https://github.com/checkstyle/checkstyle/issues/3744#issuecomment-273989531)
* An additional configuration (in some way, property or command-line argument), off by default, to do the replacement.
* Some other idea.

**Comment by rnveach** [comment]
@snago Unfortunately the issue has not been approved by the other admin, so it hasn't been agreed on what the fix will be.

**Comment by romani** [comment]
solution provided at https://github.com/checkstyle/checkstyle/issues/3744#issuecomment-273989531 is pretty good and backward compatible to existing config files (that is awesome). Fortunately in world we have only two separator type we need to support.

The only problem I see is then Windows user write suppression in windows style(that is highly expected) then Linux user will be in trouble.
Here is cut from  @viliam-durina  example:
```
	file=c:\work\Program.java
	regexp=\\work\\Program.java
	fileMatches=true


	file=/work/Program.java
	regexp=\\work\\Program.java
	fileMatches=false
	-----------
```
so problem will always be, but fortunately it is easy to fix, search `\\` replace with `/`
```
	regexp=\\work\\Program.java

	regexp=/work/Program.java
```
If suppression file already exists, and working in multi OS mode, there will be examples already, so users will use existing suppression as example.


Note:
1) usage of `treatAsFileSystemSerator` is more explicit/exact solution, but it will not support backward compatibility. Users will use either `\` or `/`. So attempt to try them both looks reasonable. 
2) usage of special attribute that controls this (like by Ant's `handledirsep`) could be done later on, if some users complain and explain in what context it is better control this. But introduction of new attribute will be backward compatible.

==================

I am voting to approve  @viliam-durina  solution to be applied.

@rnveach  and @pbludov , please review and if no red flags ... lets put a label "approved" on issue and let @snago to make it happen.

**Comment by rnveach** [comment]
> `containsRegexp(regexp, file.replace('\\', '/'))`

You can't parse a regexp like this.
If users write `\.` to say period instead of any single character, replacement will break the regular expression as now `\.` (match period) will become `/.` (linux path plus 1 character).

**Comment by romani** [comment]
Agree.

Name could be `convertFilePathToForwardSlashedOnMismatch` but it is very long so result config will ugly, even more ugly than with `[\\/]`. Examples:
```
<suppress checks="OuterTypeFilename"
  files="[\\/]package-info\.java"/>

<!--  is it look better -->
<suppress checks="OuterTypeFilename" convertFilePathToForwardSlashedOnMismatch="true"
  files="/package-info\.java"/>
<suppress checks="OuterTypeFilename" standardizeFileSeparator="true"
  files="/package-info\.java"/>
```

```
<suppress checks="PackageDeclarationCheck"
  files="[\\/]resources[\\/].*[\\/]singlespaceseparator[\\/]InputSingleSpaceSeparatorEmpty\.java"/>

  <suppress checks="PackageDeclarationCheck"
files="[\\/]resources[\\/].*[\\/]singlespaceseparator[\\/]InputSingleSpaceSeparatorEmpty\.java"/>
```
or short `fileSeparator`=`/`

**Comment by rnveach** [comment]
This logic would have to be added to all filters, not just suppression, that have files. This includes xpath and the single filter.

**Comment by snago** [comment]
One idea is to put this property on the suppressions-tag to set it once for all the included suppress tags. This would mean less clutter on each suppress line. E.g.:
```xml
<suppressions convertFilePathToForwardSlashedOnMismatch="true">
    <suppress checks="OuterTypeFilename" files="..." />
    <suppress checks="PackageDeclarationCheck" files="..." />
</suppressions>
```
----
> Please share your suppression configs, may be you have different view.

So far our suppression config is pretty simple:
```xml
<suppressions>
    <!-- Suppress naming checks for autogenerated files. -->
    <suppress files="[/\\]jaxb[/\\]" checks="ConstantName|MemberName"/>
    <suppress files="[/\\]protobuf[/\\]" checks="ConstantName|LocalVariableName|MemberName"/>
</suppressions>
```
IMO when reading `files="[/\\]jaxb[/\\]"` it is not clear that `[/\\]` is a hack to get it to work on both Windows and Linux, and not some special checkstyle syntax.

**Comment by viliam-durina** [comment]
People tend to use native path separators of their platform. Unless somebody tries to build on a platform with different separator, they won't realize. When they do, the first thing they do is they go to google. And there comes the _what-the-heck moment_ when they find out they need to use the `[\\/]` to be portable. This was my story.

Using `standardizeFileSeparator` in _every single_ rule is cumbersome too. I'd be fine with a global option, not a per-rule one.

I vote for either of these two:
- my proposal without a global config, that is `/` will be a portable path separator, `\` will only work on windows
- completely ignoring whether `/` or `\` is used, with a _global_ option to enable it. That is both `/` and `\` will work on all OSes

**Comment by romani** [comment]
Lets wrap up discussion....
Looks like extra attribute is more preferable as it could allow users to avoid extra regexp match (some small performance benefit). If user do not care about performance (usual case when suppression files are not big) he can use option and enjoy.

Lets vote on name
`<suppressions convertFilePathToForwardSlashedOnMismatch="true">`
vs
`<suppressions standardizeFileSeparator="true">`

please share votes.

**Comment by snago** [comment]
My vote is on `convertFilePathToForwardSlashedOnMismatch`. It's a bit long but explains what it does very well.

I would prefer if the default was the other way around though. So that by default "it just works" and those who really need to squeeze out maximum performance can turn it off and deal with path separator issues how they see fit.

Otherwise this would continue to be a common pitfall to encounter when starting out with checkstyle and suppressions.

But starting out it can be off by default. Then when it's implemented and released it's easier to get performance comparisons done on different projects to see how much of an impact it would cause to have it on by default.

**Comment by pbludov** [comment]
I vote for `<suppressions standardizeFileSeparator="true">`.

I like the idea of not naming and distinguishing separators, compared to the ideas of turning "the other" into "the right".

**Comment by romani** [comment]
>standardizeFileSeparator

By name it looks like converting file path first, and then do all operations.
Backslash is escape symbol in regexp, in an accurate regexp we will convert it something even more weird. I do not know term "standard file separator".

@strkkk , @viliam-durina , please share your votes on name and default value.

**Comment by pbludov** [comment]
>I do not know term "standard file separator".

`Path` separator?

**Comment by romani** [comment]
It is different, it is for separation between paths (`:` Linux  , `;` windows)

**Comment by viliam-durina** [comment]
As we're are adding an option, we're free to do whatever. You were worried about the performance of evaluating the regex twice. We can add the option `forceLinuxDirectorySeparator` which will always replace backslashes with forward slashes, if `File.separatorChar == '\\'`. There regexp will only work if `/` is used. If `\\` is used, it will not work even on windows. What do you think? Also all the samples should be written with the option enabled and `/` used everywhere.

Also Java uses the term "name separator" in `File.separator`. This IMO is too generic. "Path separator" is confused with separating the paths, "file separator" feels wrong, because we don't _separate files_. Therefore I used "directory separator".

**Comment by romani** [comment]
> We can add the option forceLinuxDirectorySeparator which will always replace backslashes with forward slashes

I  ok with such approach, as in most cases user define relative path (so "c:\" is very bad practice).
better name is `forceLinuxDirectorySeparator`.

> Therefore I used "directory separator".

it the same as "path separator" for me, not good.

@rnveach , please share your thoughts on this. Are you ok with `forceLinuxFileSeparator` ? so path with violation will be converted by regexp ones, and then be matched to suppression regexp.

I still do not believe that there is performance issue with regexp, with `forceLinuxFileSeparator` on projects with a lot violation and big supporession file (I do not know such projects), regexp will not be a problem, but creation of string copies of converted paths might be. As we can not avoid seconds regexp execution, as you still have concern, I would do `convertFilePathToForwardSlashedOnMismatch` so at least be lazy a bit for user who sit the same OS.

To finish with discussion quicker, I propose to go with `convertFilePathToForwardSlashedOnMismatch` enabled by default. During implementation we will ask contributir to test performance, to make decision on activation of it by default or not. Or even more, remove this attribute, and always to convert as originally requested.

@rnveach , deal ?

**Comment by rnveach** [comment]
I am sorry. I thought I replied to this but really I just forgot about it.

> forceLinuxFileSeparator 

I am fine with this.

> convertFilePathToForwardSlashedOnMismatch
> implementation we will ask contributir to test performance

I am fine with this.

**Comment by rnveach** [comment]
Anyone is free to start on issue. Just post "I am on it" so we don't have a race.

**Comment by snago** [comment]
I'll take a swing at creating a PR.

**Comment by romani** [comment]
issue is approved at https://github.com/checkstyle/checkstyle/issues/3744#issuecomment-508875306  we waiting (anybody in world is welcome) for implementation to be contributed to project.

**Comment by romani** [comment]
>I think you had performance concerns with this?

I doubt it will be visible performance degradation.  But we will do some validation as we see implementation. Please send PR and we will guide you on a process of acceptance.

**Comment by romani** [comment]
fix is merged

temp docs:
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/49fd839_20260721033945/filters/suppressionfilter.html#Notes example is also udpated
https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/49fd839_20260721033945/filters/suppressionxpathfilter.html#Notes
